compiler/rustc_hir_typeck/src/pat.rs

```diff
@@ -629,12 +629,20 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         expected: Ty<'tcx>,
         pat_info: PatInfo<'tcx, '_>,
     ) -> Ty<'tcx> {
-        let PatInfo { binding_mode: def_bm, top_info: ti, .. } = pat_info;
+        let PatInfo { binding_mode: BindingAnnotation(def_br, _), top_info: ti, .. } = pat_info;
 
         // Determine the binding mode...
         let bm = match ba {
-            BindingAnnotation(ByRef::No, Mutability::Not) => def_bm,
-            _ => ba,
+            BindingAnnotation(ByRef::No, Mutability::Mut)
+                if !(pat.span.at_least_rust_2024()
+                    && self.tcx.features().mut_dont_reset_binding_mode_2024)
+                    && matches!(def_br, ByRef::Yes(_)) =>
+            {
+                // `mut x` resets the binding mode in edition <= 2021.
+                BindingAnnotation(ByRef::No, Mutability::Mut)
+            }
+            BindingAnnotation(ByRef::No, mutbl) => BindingAnnotation(def_br, mutbl),
+            BindingAnnotation(ByRef::Yes(_), _) => ba,
```

Comment:
This doesn't preserve old behavior in the case where `ba` is `BindingAnnotation(ByRef::No, Mutability::Not)` and `def_bm.1` is `Mutability::Mut`. I guess that can't happen, right? Could we store just a `ByRef` in `PatInfo`?

Reply:
> Could we store just a `ByRef` in `PatInfo`?

I think we can, but it would require changes to `calc_default_binding_mode`, so I'll do it in a later PR so as not to conflict with #123512